libs/arcade-core/arcade_core/executor.py

```diff
@@ -59,7 +59,7 @@ async def run(
             if asyncio.iscoroutinefunction(func):
                 results = await func(**func_args)
             else:
-                results = func(**func_args)
+                results = await asyncio.to_thread(func, **func_args)
```

Comment:
(16 month old code) Sync tools were running directly on the uvicorn event loop preventing other tool calls from being processed.

I could see `asyncio.to_thread` causing some issues with the GIL if there ends up being multiple synchronous tools being executed this way simultaneously. But I think it's still better than completely blocking all other tool calls? @evantahler @sdreyer

Reply:
Something else you could try, is set uvicorn works to more than 1. I believe that makes them not share a single GIL

Reply:
I don't _think_ `uvicorn.run(..., workers=4)` will work for MCP sessions (because MCP sessions would then be process specific) and then `MCPApp`'s runtime capabilities would break since they would only affect one of the N processes so we would end up with uvicorn workers that have different tool catalogues, prompts, and resources

Reply:
So I looked into what the default number of 'workers' for Python's `ThreadPoolExecutor`:

- Python >=3.8,<3.13: `min(32, os.cpu_count() + 4)` or in other words, 5 in production.
- Python >=3.13: `min(32, (os.process_cpu_count() or 1) + 4`, or in other words, 5 in production.

[Source](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor)

Reply:
So I looked into what the default number of 'workers' for Python's `ThreadPoolExecutor`:

- Python >=3.8,<3.13: `min(32, os.cpu_count() + 4)` or in other words, 5 in production.
- Python >=3.13: `min(32, (os.process_cpu_count() or 1) + 4`, or in other words, 5 in production.

[Source](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor)